### PR TITLE
Improve error logging; Add drag and drop events

### PR DESCRIPTION
**Pull request by ScraM-Team**
In a large project, it is hard to tell the source of errors in Flavour output, since filenames are missing in some cases.

This change provides more detail in error logs.  Now HTML errors from jericho, and EL parsing errors include a filename.  Example of the new output:

> [INFO] Running TeaVM
> ERROR: **teavm.errors.Client1**: StartTag at (r3,c3,p49) missing required end tag
> [INFO] Output file built with errors
> [INFO] Debug information successfully written
> [INFO] Source maps successfully written
> [INFO] Source files successfully written
> [ERROR] **templates/client2.html**: Property not found: noSuchField

This change also includes additional event types.

### CODE REVIEW

File: templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java [hunk 1]
```diff
@@ -91,7 +91,7 @@ private TemplateInfo parseForModel(ReflectClass<Object> cls, SourceLocation loca
             for (Diagnostic diagnostic : parser.getDiagnostics()) {
                 SourceLocation diagnosticLocation = location != null ? new SourceLocation(location.getMethod(), path,
                         mapper.getLine(diagnostic.getStart()) + 1) : null;
-                diagnostics.error(diagnosticLocation, diagnostic.getMessage());
+                diagnostics.error(diagnosticLocation, path + ": " + diagnostic.getMessage());
```
Comment (konsoletyper) [inline templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java]: Why this change at all? Diagnostics already takes location (including file name) with location.

File: templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java [hunk 2]
```diff
@@ -114,6 +115,47 @@ public boolean wasSuccessful() {
 
     public List<TemplateNode> parse(Reader reader, String className) throws IOException {
         source = new Source(reader);
+        source.setLogger(new Logger() {
```
Comment (konsoletyper) [inline templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java]: I don't think it's the right way. There's TeaVM diagnostics which can be used to report errors and warnings. Also, I doubt if it's really necessary to report debug and info messages.

File: templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java (hunk 1 above)
Comment (ScraM-Team) [reply templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java]: OK, let me see if I can add the file name to the diagnostic.  No filename is printed when using the 'master' branch code, so maybe it is null right now.

File: templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java (hunk 2 above)
Comment (ScraM-Team) [reply templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java]: OK, I will try adding filenames to the diagnostic.  I'll update the PR when it is ready.

File: templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java (hunk 1 above)
Comment (ScraM-Team) [reply templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java]: For "Property not found", location is null (from TemplatingProxyGenerator:42, which calls Metaprogramming.getLocation(), which apparently returns null).  If location is null, then SourceLocation is null on line 94 and the filename won't be printed

If I try to construct a SourceLocation I don't have a ReflectMethod, do you have any recommendation?

File: templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java (hunk 2 above)
Comment (ScraM-Team) [reply templates/src/main/java/org/teavm/flavour/templates/parsing/Parser.java]: I couldn't figure out how to extract a list of errors or warnings from Jericho.  I did figure out a way to preserve the original logger log level so it won't report debug and info messages.  Updated PR coming shortly for your review.

File: templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java (hunk 1 above)
Comment (ScraM-Team) [reply templates/src/main/java/org/teavm/flavour/templates/emitting/TemplatingProxyGenerator.java]: In my updated PR it only adds the path if the location is null.  Let me know if there is a better way.

**Comment by ScraM-Team** [comment]
I will close this and reopen it as 2 separate PRs against the current master.
